[JSC] uDFG should be able to watch JSGlobalObject WatchpointSets

Constellation opened the pull request:
#### 085b9807d093d97e6d2df08c98904a85e09465a6
<pre>
[JSC] uDFG should be able to watch JSGlobalObject WatchpointSets
<a href="https://bugs.webkit.org/show_bug.cgi?id=247154">https://bugs.webkit.org/show_bug.cgi?id=247154</a>
rdar://101657085

Reviewed by Mark Lam.

This patch allows uDFG to watch JSGlobalObject WatchpointSets.
We extract this watchpoint insertion as a LinkerIR. And we run this when materializing DFG::JITData,
which should eventually become uDFG&apos;s linking phase. We also fixed a bug where we missed visiting
uDFG CallLinkInfo in CodeBlock.

* Source/JavaScriptCore/bytecode/CodeBlock.cpp:
(JSC::CodeBlock::finalizeJITInlineCaches):
(JSC::CodeBlock::getICStatusMap):
(JSC::CodeBlock::getCallLinkInfoForBytecodeIndex):
* Source/JavaScriptCore/dfg/DFGGraph.h:
* Source/JavaScriptCore/dfg/DFGJITCode.cpp:
(JSC::DFG::JITData::JITData):
(JSC::DFG::attemptToWatch):
(JSC::DFG::JITData::tryInitialize):
* Source/JavaScriptCore/dfg/DFGJITCode.h:
(JSC::DFG::JITData::tryCreate):
(JSC::DFG::JITData::create): Deleted.
* Source/JavaScriptCore/dfg/DFGJITCompiler.cpp:
(JSC::DFG::JITCompiler::link):
(JSC::DFG::JITCompiler::addToConstantPool):
* Source/JavaScriptCore/dfg/DFGJITCompiler.h:
* Source/JavaScriptCore/dfg/DFGJITFinalizer.cpp:
(JSC::DFG::JITFinalizer::finalize):
* Source/JavaScriptCore/dfg/DFGPlan.cpp:
(JSC::DFG::Plan::tryFinalizeJITData):
(JSC::DFG::Plan::finalizeJITData): Deleted.
* Source/JavaScriptCore/dfg/DFGPlan.h:

Canonical link: <a href="https://commits.webkit.org/256111@main">https://commits.webkit.org/256111@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2819875e278acdb3476d8eefe341093942fc373d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94733 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/3889 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/27611 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/104352 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/164616 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/3957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/32078 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/87045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/100282 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100400 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/3957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/81093 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/87045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/3957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/27611 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/87045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/85904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/38483 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/27611 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/38/builds/81107 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/36327 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/27611 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/38/builds/81107 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/63/builds/40240 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/81093 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/83776 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2012 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/65/builds/42212 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/27611 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/45/builds/18917 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/dfg/DFGGraph.h:
```diff
@@ -816,112 +830,79 @@ class Graph final : public virtual Scannable {
         JSGlobalObject* globalObject = globalObjectFor(node->origin.semantic);
         InlineWatchpointSet& set = globalObject->arrayPrototypeChainIsSaneWatchpointSet();
-        return isWatchingGlobalObjectWatchpoint(globalObject, set);
+        return isWatchingGlobalObjectWatchpoint(globalObject, set, LinkerIR::Type::ArrayPrototypeIsSaneChainWatchpointSet);
```
Why name this `ArrayPrototypeIsSaneChainWatchpointSet` instead of `ArrayPrototypeChainIsSaneWatchpointSet` to match `arrayPrototypeChainIsSaneWatchpointSet()`?

Constellation replied:
Changed

MenloDorian commented on Source/JavaScriptCore/dfg/DFGJITCode.cpp:
```diff
@@ -35,12 +35,60 @@
     : Base(jitCode.m_linkerIR.size())
     , m_stubInfos(jitCode.m_unlinkedStubInfos.size())
     , m_callLinkInfos(jitCode.m_unlinkedCallLinkInfos.size())
+    , m_watchpoints()
```
This is not needed.  Please remove.

Constellation replied:
Removed

MenloDorian commented on Source/JavaScriptCore/dfg/DFGJITCode.cpp:
```diff
@@ -64,11 +112,81 @@ JITData::JITData(VM& vm, CodeBlock* codeBlock, const JITCode& jitCode, ExitVecto
         }
-        default:
+        case LinkerIR::Type::HavingABadTimeWatchpointSet: {
+            auto* globalObject = codeBlock->globalObject();
```
Would it be worth pre-caching this `globalObject` outside of this loop?

Constellation replied:
I think we should not do that since in the future we will extend it to accept different JSGlobalObject.

MenloDorian commented on Source/JavaScriptCore/dfg/DFGJITCode.cpp:
```diff
@@ -64,11 +112,81 @@ JITData::JITData(VM& vm, CodeBlock* codeBlock, const JITCode& jitCode, ExitVecto
+        case LinkerIR::Type::HavingABadTimeWatchpointSet: {
+            auto* globalObject = codeBlock->globalObject();
+            auto& watchpoint = m_watchpoints[indexOfWatchpoints++];
+            failure |= !attemptToWatch(codeBlock, globalObject->havingABadTimeWatchpointSet(), watchpoint);
```
I think it'll be clearer if you use a `success` bool instead of `failure` and express this as:
```
success &= attemptToWatch(codeBlock, globalObject->havingABadTimeWatchpointSet(), watchpoint);
```

Ditto for below.

Constellation replied:
Changed.

MenloDorian commented on Source/JavaScriptCore/dfg/DFGGraph.h:
```diff
@@ -816,112 +830,79 @@ class Graph final : public virtual Scannable {
         JSGlobalObject* globalObject = globalObjectFor(node->origin.semantic);
         InlineWatchpointSet& set = globalObject->arrayPrototypeChainIsSaneWatchpointSet();
-        return isWatchingGlobalObjectWatchpoint(globalObject, set);
+        return isWatchingGlobalObjectWatchpoint(globalObject, set, LinkerIR::Type::ArrayPrototypeIsSaneChainWatchpointSet);
```
I suggest naming `ArrayPrototypeIsSaneChainWatchpointSet` as `ArrayPrototypeChainIsSaneWatchpointSet` to be consistent with the function name `arrayPrototypeChainIsSaneWatchpointSet()`.

Constellation replied:
Changed.

MenloDorian commented on Source/JavaScriptCore/dfg/DFGGraph.h:
```diff
@@ -816,112 +830,79 @@ class Graph final : public virtual Scannable {
         JSGlobalObject* globalObject = globalObjectFor(node->origin.semantic);
         InlineWatchpointSet& set = globalObject->stringPrototypeChainIsSaneWatchpointSet();
-        return isWatchingGlobalObjectWatchpoint(globalObject, set);
+        return isWatchingGlobalObjectWatchpoint(globalObject, set, LinkerIR::Type::StringPrototypeIsSaneChainWatchpointSet);
```
Ditto.

Constellation replied:
Changed.

MenloDorian commented on Source/JavaScriptCore/dfg/DFGGraph.h:
```diff
@@ -816,112 +830,79 @@ class Graph final : public virtual Scannable {
         JSGlobalObject* globalObject = globalObjectFor(node->origin.semantic);
         InlineWatchpointSet& set = globalObject->objectPrototypeChainIsSaneWatchpointSet();
-        return isWatchingGlobalObjectWatchpoint(globalObject, set);
+        return isWatchingGlobalObjectWatchpoint(globalObject, set, LinkerIR::Type::ObjectPrototypeIsSaneChainWatchpointSet);
```
Ditto.

Constellation replied:
Changed.

MenloDorian reviewed (approved):
r=me with suggestions for improvement.

Constellation committed:
[JSC] uDFG should be able to watch JSGlobalObject WatchpointSets
https://bugs.webkit.org/show_bug.cgi?id=247154
rdar://101657085

Reviewed by Mark Lam.

This patch allows uDFG to watch JSGlobalObject WatchpointSets.
We extract this watchpoint insertion as a LinkerIR. And we run this when materializing DFG::JITData,
which should eventually become uDFG's linking phase. We also fixed a bug where we missed visiting
uDFG CallLinkInfo in CodeBlock.

* Source/JavaScriptCore/bytecode/CodeBlock.cpp:
(JSC::CodeBlock::finalizeJITInlineCaches):
(JSC::CodeBlock::getICStatusMap):
(JSC::CodeBlock::getCallLinkInfoForBytecodeIndex):
* Source/JavaScriptCore/dfg/DFGGraph.h:
* Source/JavaScriptCore/dfg/DFGJITCode.cpp:
(JSC::DFG::JITData::JITData):
(JSC::DFG::attemptToWatch):
(JSC::DFG::JITData::tryInitialize):
* Source/JavaScriptCore/dfg/DFGJITCode.h:
(JSC::DFG::JITData::tryCreate):
(JSC::DFG::JITData::create): Deleted.
* Source/JavaScriptCore/dfg/DFGJITCompiler.cpp:
(JSC::DFG::JITCompiler::link):
(JSC::DFG::JITCompiler::addToConstantPool):
* Source/JavaScriptCore/dfg/DFGJITCompiler.h:
* Source/JavaScriptCore/dfg/DFGJITFinalizer.cpp:
(JSC::DFG::JITFinalizer::finalize):
* Source/JavaScriptCore/dfg/DFGPlan.cpp:
(JSC::DFG::Plan::tryFinalizeJITData):
(JSC::DFG::Plan::finalizeJITData): Deleted.
* Source/JavaScriptCore/dfg/DFGPlan.h:

Canonical link: https://commits.webkit.org/256111@main
